sys/riscv/riscv/pmap.c

```diff
@@ -2419,14 +2430,29 @@ pmap_fault_fixup(pmap_t pmap, vm_offset_t va, vm_prot_t ftype)
 
 	if ((pmap != kernel_pmap && (oldpte & PTE_U) == 0) ||
 	    (ftype == VM_PROT_WRITE && (oldpte & PTE_W) == 0) ||
+	    (ftype == VM_PROT_WRITE_CAP && (oldpte & PTE_W) == 0) ||
 	    (ftype == VM_PROT_EXECUTE && (oldpte & PTE_X) == 0) ||
 	    (ftype == VM_PROT_READ && (oldpte & PTE_R) == 0))
 		goto done;
 
 	bits = PTE_A;
-	if (ftype == VM_PROT_WRITE)
+	if ((ftype == VM_PROT_WRITE) || (ftype == VM_PROT_WRITE_CAP))
 		bits |= PTE_D;
 
+	if (ftype == VM_PROT_WRITE_CAP) {
+		/*
+		 * XXX until hardware's caught up, we sometimes set PTE_CD
+		 * and leave PTE_SC clear to force traps.  If PTE_CD is
+		 * clear, we don't mean for this page to be cap-capable;
+		 * eventually this will be expressed directly in PTE_SC,
+		 * at which point this should be setting PTE_CD.
+		 */
```

Comment:
This says what you do, not why you do it, nor what hardware needs to do to "catch up".